### ISSUE TITLE
Singular Matrix Error in 04-dmet-nio-fm Example

### ISSUE THREAD

**Issue by chemsm** [gkclab/libdmet_preview#19]
Hello, I am a researcher trying to replicate the 04-dmet-nio-fm problem example provided in your library as I explore using DMET for my studies. 

While I was able to successfully replicate the example 03-dmet-nio-afm, running the code for the 04-dmet-nio-fm example as provided leads to a UICCSD calculation error displaying the following message:

******* WARN: diis singular, eigh(h) [ -0.0001 -0. 0. 0. 202.2266 32164.7754 13988017.9382]
*******

This causes the computation to halt. The traceback is as follows:

Traceback (most recent call last):
  File "/home/chemsm/results/nio_fm.py", line 282, in <module>
    dmet.SolveImpHam_with_fitting(Lat, Filling, ImpHam, basis, solver, \
  File "/home/chemsm/programs/libdmet_preview/libdmet/dmet/Hubbard.py", line 195, in __call__
    rhoEmb, EnergyEmb, nelec = solve_with_mu_loop(0.0)
                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/libdmet_preview/libdmet/dmet/Hubbard.py", line 141, in solve_with_mu_loop
    SolveImpHam_with_dmu(lattice_I, ImpHam_I, basis_I, mu, 
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/libdmet_preview/libdmet/dmet/Hubbard.py", line 87, in SolveImpHam_with_dmu
    result = solver.run(ImpHam, **solver_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/libdmet_preview/libdmet/solver/cc.py", line 302, in run
    E_corr, t1, t2 = self.cisolver.kernel(t1=t1, t2=t2, eris=eris)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/cc/uccsd.py", line 598, in kernel
    return self.ccsd(t1, t2, eris, mbpt2)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/cc/uccsd.py", line 614, in ccsd
    return ccsd.CCSDBase.ccsd(self, t1, t2, eris)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/cc/ccsd.py", line 1083, in ccsd
    kernel(self, eris, t1, t2, max_cycle=self.max_cycle,
  File "/home/chemsm/programs/pyscf/pyscf/cc/ccsd.py", line 84, in kernel
    t1, t2 = mycc.run_diis(t1, t2, istep, normt, eccsd-eold, adiis)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/cc/ccsd.py", line 1195, in run_diis
    t1, t2 = self.vector_to_amplitudes(adiis.update(vec))
                                       ^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/lib/diis.py", line 237, in update
    self._xprev = xnew = self.extrapolate(nd)
                         ^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/pyscf/pyscf/lib/diis.py", line 264, in extrapolate
    raise e
  File "/home/chemsm/programs/pyscf/pyscf/lib/diis.py", line 261, in extrapolate
    c = numpy.linalg.solve(h, g)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/anaconda3/lib/python3.11/site-packages/numpy/linalg/linalg.py", line 409, in solve
    r = gufunc(a, b, signature=signature, extobj=extobj)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chemsm/programs/anaconda3/lib/python3.11/site-packages/numpy/linalg/linalg.py", line 112, in _raise_linalgerror_singular
    **raise LinAlgError("Singular matrix")
numpy.linalg.LinAlgError: Singular matrix**

**Comment by zhcui** [comment]
Hello chemsm,

Thanks for reporting this.

From the output, the UHF-in-UHF embedding is exact. However, the UCCSD based on the UHF looks weird. The t amplitudes becomes diverge after a few CC iterations.
This example has been tested before (with pyscf 2.0) and gives typical results. I will try to reproduce your error based on the latest version.

I will get back to you after I re-run it.

**Comment by chemsm** [comment]
Thank you for your reply. Since I am unable to fix the code on my own, I will await further communication from the developer regarding updates or improvements that will render the code executable in the latest version of PySCF.
Thank you,
chemsm

**Comment by zhcui** [comment]
Hi chemsm,

I rerun the example scripts.

However, I can't reproduce your error when using the latest version of pyscf.
Here is my output:
[03-dmet-nio-afm.txt](https://github.com/gkclab/libdmet_preview/files/14901320/03-dmet-nio-afm.txt)

It looks like we can get the same periodic HF solution, and in the CC solver, HF energy is the same. However, MP2 energy is different. This is strange. Can you try to use the latest version of pyscf?

**Comment by chemsm** [comment]
Thank you for rerunning the example scripts. I also tried running my previous code(That is made a little modification where lines[877] = 'class CIAHMinimizer(ciah.CIAHOptimizerMixin):\n'.) again.
Additionally, I've been using pyscf version 2.5.0 all along.

I also tried running the code exactly as in your 03-dmet-nio-afm.txt output file, but I'm still getting the same singular matrix error. 

Interestingly, when I download and run the nio_afm DMET code directly from its directory, I encounter an error that says "array must not contain infs or NaNs". I have attached the relevant files below.

In a nutshell,
I rerun
1.  initial error code (a little modification ver) -> singular matrix error
2.  your successful recent attached code  -> singular matrix error
3.  03-afm original code & original path -> valueError: array must not contain infs or NaNs
4.  04-fm original code & original path -> singular matrix error

I attached 3, 4 output file.
[nio_afm.txt](https://github.com/gkclab/libdmet_preview/files/14961533/nio_afm.txt)
[nio_fm.txt](https://github.com/gkclab/libdmet_preview/files/14961534/nio_fm.txt)
